### DISCUSSION TITLE
deploy on unraid?

### DISCUSSION THREAD

**Discussion by tmyflyte** [mediacms-io/mediacms#331] [Feature Requests]
Hello, I'm an unraid user and I'm having a hard time deploying it there. Maybe you could make a container that is natively compatible with unraid?

Some guidelines can be found here - https://github.com/selfhosters/unRAID-CA-templates (especially the guide from Selfhosters.net)

Thanks and I hope we can sort something out :)

**Comment by dsfsdbvxc** [comment]
+1 for this!

**Comment by Jtav1** [comment]
Hey I just got it mostly working and put a brief write-up on my personal blog. If I remember I will come back and link it when I have a few kinks worked out (I'm here to learn how to fix a few things)

**Reply by Jtav1** [reply]
@tmyflyte Ok so I was able to work out a lot of the kinks. This isn't going to be a basic Unraid template install because that's not how templates work, but there is a good docker-compose UI plugin that allows you to run container stacks in a way that is compatible with the install instructions here. 

![image](https://user-images.githubusercontent.com/5510644/185724113-45ae6e4b-aff3-464d-bd3d-0ea0746ed35b.png)


As promised, I did write blog posts about how I got it working so far. I do have some more work to do but this might be helpful. Link to the first article below, and the full text in case you can't get to my site (blog is selfhosted, no monetization or ads or accounts or anything, fully public)

https://blog.justintavares.com/selfhosted-youtube-alternative-on-unraid/

<details><summary>Full text (click to expand)</summary>
I finally decided to look for some good video sharing systems that I could selfhost and I found [MediaCMS](https://mediacms.io/) to be the shiniest and prettiest. It's linked on the [Awesome-Selfhosted list](https://github.com/awesome-selfhosted/awesome-selfhosted#media-streaming---video-streaming), and is open source. There are alternatives on that list that may better suit your needs, one I also considered was [Peertube ](https://joinpeertube.org/)but I'm not fully happy with how it looks. I just got MediaCMS working, so I may later decide to switch to another solution but for now lets look at how to get it set up on Unraid.

I like Docker. Fortunately, MediaCMS comes with [Docker install instructions](https://github.com/mediacms-io/mediacms/blob/main/docs/admins_docs.md#3-docker-installation)! But for... docker-compose. Which Unraid doesn't come with a UI for, despite it being accessible in terminal. If you're like me, terminal stuff is annoying to you, especially remote terminal stuff where you have to send files over and then do things with them. But because the Unraid community is fantastic, there's a WebUI plugin for docker compose - see [Ibracorp's great article on it here](https://docs.ibracorp.io/docker-compose/docker-compose-for-unraid).
Docker-compose configuration for your server

Once the docker-compose UI is set up, you can just set up a new docker-compose "stack" and then edit the configuration and paste your selected docker-compose file contents into it. For MediaCMS I went with the [standard docker-compose](https://github.com/mediacms-io/mediacms/blob/main/docker-compose.yaml), but there are [other options](https://github.com/mediacms-io/mediacms/blob/main/docs/admins_docs.md#4-docker-deployment-options) if you want to include other components. I have the other necessary components already taken care of with other containers, and did not need those configurations.
Volume example

Next, you want to update your docker-compose file to work in your environment. I want my files to exist in externally-mounted directories just like the rest of my apps in my appdata directory so the first thing I did was go through and replace all of the volume locations with the external volume /mnt/user/appdata/MediaCMS/

I'm still figuring out how to move the media storage location. I'll figure it out!

Note that all the services seem to want to live in the same directory, except the database (db) container that I would suggest setting to a separate one just for cleanliness. I mounted the db container's volume to appdata/MediaCMS-db/
Port for the web container

Set the port on the web container to an unused Unraid port. No custom network configuration is specified, so the web container needs a host port to listen to from its own default port 80.

Now on the migrations container, you want to uncomment out the ADMIN_PASSWORD environment variable and add your own. This is the password for the admin login on the site, so make a good one.

You don't have to change any container names, the Unraid docker-compose plugin will preface them with the stack name on your container list. Once you're finished with the above, make sure you hit Save Changes. Then hit the Compose Pull button to tell your system to pull down all the images as needed. Once that's finished, hit Compose Up to run all the containers, or Compose Down to shut them all off. If you've used docker-compose before those commands should be familiar.

But before you hit that Compose Up button, there's one more step... The part that I didn't immediately notice is that MediaCMS assumes you have the website's code in the externally mounted volume. As far as I can tell, all you have to do is download or check out the [root of the code repo on github](https://github.com/mediacms-io/mediacms) into your appdata/MediaCMS/ directory.
Setting up and using your MediaCMS instance

This isn't terribly confusing and is semi covered in the github [configuration ](https://github.com/mediacms-io/mediacms/blob/main/docs/admins_docs.md#5-configuration)docs, so I'll be short.

I pointed a reverse proxy to my site before continuing, but that's not necessary for internal LAN-only use and is a separate topic.

In your appdata/MediaCMS/ directory (you may have to give yourself permissions via the Unraid terminal if you access this from another machine), the cms/ directory contains a settings.py file. This file contains a bunch of config settings for the site. To override them, you copy pieces out of the settings.py file and insert them into the following file: /MediaCMS/deploy/docker/local_settings.py

I will dive into these later and cover it in a future blog post, as well as some other things like moving your media storage to another location.

That's all I have for now, comment if you have any questions and maybe I can look into them too!
</details>

**Reply by tmyflyte** [reply]
that is awesome @Jtav1  !!!

I'm out this weekend but will def get hands-on by the beginning of the week

will make sure to post back with comments :)

thanks!!!

**Reply by tmyflyte** [reply]
So I followed the posts but somehow mine isn't working yet. I'll have to fiddle with it later again but this is how it looks right now :(
![image](https://user-images.githubusercontent.com/24412030/186774759-7ce44d5c-8892-40fc-9839-cccb3704e68a.png)

**Reply by Jtav1** [reply]
Definitely check the logs on the individual containers but I think those all use the same image so it might be an issue with your pull? Not sure. If nothing useful is in the logs maybe post your (redacted) docker compose

**Reply by tmyflyte** [reply]
my docker-compose.yaml - https://pastebin.com/PADiFLwf
my local_settings.py - https://pastebin.com/dc6NYYEB
log is only partcial - https://pastebin.com/4yErHeNx

this is the first error I see

![image](https://user-images.githubusercontent.com/24412030/187044302-e364f152-1ba7-4be5-9571-da1aebe052b9.png)

I have no idea what could be wrong in entrypoint.sh, I never touched it 

something else I noticed is that ports for redis and db were already taken so I changed below in local_settings.py, still no luck though

```
DATABASES = {
    "default": {
        "ENGINE": "django.db.backends.postgresql",
        "NAME": "mediacms",
        "HOST": "127.0.0.1",
        "PORT": "5331",
        "USER": "mediacms",
        "PASSWORD": "mediacms",
    }
}

REDIS_LOCATION = "redis://127.0.0.1:6339/1"
```

**Reply by Jtav1** [reply]
I'm assuming the user in the container doesn't have permission to execute anything in the directory with your code. You may have to go in the container console or the unraid server console and chmod the files in that directory to the right permissions for your docker container to run them. 

"redis" and "db" names can be taken, that's fine, I'd move them back to the default names. The unraid plugin will preface the container names with the "stack" name which is common with docker compose stacks. e.g. it named mine mediacms-redis-1, etc 

![image](https://user-images.githubusercontent.com/5510644/187046871-2ab3261f-93cf-4ee6-9917-13a41c34ac1e.png)

If that helps, please let me know and I'll add it to my instructions above/in the article too.

**Reply by tmyflyte** [reply]
> "redis" and "db" names can be taken, that's fine, I'd move them back to the default names. The unraid plugin will preface the container names with the "stack" name which is common with docker compose stacks. e.g. it named mine mediacms-redis-1, etc.

No, I mean the ports!

I finally got it to work, here's how I did it: 

1. Created 'MediaCMS' stack in docker compose manager plugin
2. Edited the compose there
3. On terminal `cd /mnt/user/appdata` then `git clone https://github.com/mediacms-io/mediacms`
4. Edited /mnt/user/appdata/mediacms/deploy/docker/local_settings.py on windows
5. Back to docker compose manager, **compose pull** then **compose up**
6. All green :)

Thanks for the guidelines and tutorial though !
💯 

EDIT: my sign in page doesn't seem to be working though and I haven't been able to find out why just yet :/

![image](https://user-images.githubusercontent.com/24412030/187053407-27cd4ff4-3033-4366-b35b-e41e5ed65d2c.png)

all the log is stating is this

_[pid: 49|app: 0|req: 3/5] 129.105.186.24 () {46 vars in 3359 bytes} [Sat Aug 27 22:55:13 2022] GET /accounts/login/ => generated 145 bytes in 67 msecs (HTTP/1.1 500) 6 headers in 188 bytes (1 switches on core 0)
129.105.186.24 - - [28/Aug/2022:01:55:14 +0000] "GET /accounts/login/ HTTP/1.1" 500 145 "http://129.105.186.16:4561/" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/104.0.5112.81 Safari/537.36 Edg/104.0.1293.47"_

**Reply by Jtav1** [reply]
Heyo! Big progress, thanks I can add that to my instructions above. And sorry, didn't read you properly the first time about the ports. 

As far as the 500 goes, the system logs are located in /mediacms/logs  - I find the most useful one is debug.log

**Reply by tmyflyte** [reply]
I just opened a [new case](https://github.com/mediacms-io/mediacms/issues/509) with the details if you could also take a look - apparently it's a problem with the ports being already in use but I haven't been able to change

**Reply by Jtav1** [reply]
I don't think specifying a port is necessary, I didnt add a port for either my DB or Redis and I also have a postgresdb and redis container both listening to normal ports already. I believe they are referred to by container name, not ip/port. 

![image](https://user-images.githubusercontent.com/5510644/187133496-a941a81f-6ee1-4123-a9d6-f4ad1d91f12b.png)

Edit: See, I have other containers listening to the default ports and they aren't being touched and don't cause issues here

![image](https://user-images.githubusercontent.com/5510644/187133611-2285d0bd-ba9c-4178-8453-0f2a4fedde36.png)

**Comment by mgogoulos** [comment]
Very good article! And thanks for posting the link!

**Comment by tmyflyte** [comment]
@Jtav1 is yours like this? how do you apply the update?
If I'm not mistaken [this](https://github.com/mediacms-io/mediacms/blob/main/docs/admins_docs.md#update-1) should be similar to `cd /mnt/user/appdata` then `git clone https://github.com/mediacms-io/mediacms`? but still shows this:

![image](https://user-images.githubusercontent.com/24412030/189355658-7a9cb8cf-6fcc-49d2-bf0f-c76b40940331.png)

**Reply by Jtav1** [reply]
Mine is like that, those all use the same container. I only ever update by doing a stop -> pull -> start but it still says an update is available in that section for some reason. Not 100% sure why it does that, but I do know the dev only leaves one image tagged on docker hub (latest) so maybe the unraid UI is confused?
